cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java

```diff
@@ -127,24 +129,34 @@ public void startTx(Transaction tx) {
 
   @Override
   public List<Map<byte[], byte[]>> getPersisted(List<co.cask.cdap.api.dataset.table.Get> gets) {
-    List<Get> hbaseGets = Lists.transform(gets, new Function<co.cask.cdap.api.dataset.table.Get, Get>() {
-      @Nullable
-      @Override
-      public Get apply(co.cask.cdap.api.dataset.table.Get get) {
-        List<byte[]> cols = get.getColumns();
-        return createGet(get.getRow(), cols == null ? null : cols.toArray(new byte[cols.size()][]));
+    List<Get> hbaseGets = new ArrayList<>();
+    for (co.cask.cdap.api.dataset.table.Get get : gets) {
+      List<byte[]> cols = get.getColumns();
+      // Our Get class (co.cask.cdap.api.dataset.table.Get) with empty array means get nothing, but there is no way to
+      // specify this in an HBase Get (org.apache.hadoop.hbase.client.Get). That's why we don't call createGet for
+      // every get.
+      if (cols == null || !cols.isEmpty()) {
+        hbaseGets.add(createGet(get.getRow(), cols == null ? null : cols.toArray(new byte[cols.size()][])));
       }
-    });
+    }
+
     try {
-      Result[] results = hTable.get(hbaseGets);
-      return Lists.transform(Arrays.asList(results), new Function<Result, Map<byte[], byte[]>>() {
-        @Nullable
-        @Override
-        public Map<byte[], byte[]> apply(Result result) {
-          Map<byte[], byte[]> familyMap = result.getFamilyMap(columnFamily);
-          return familyMap != null ? familyMap : ImmutableMap.<byte[], byte[]>of();
+      Result[] hbaseResults = hTable.get(hbaseGets);
+
+      List<Map<byte[], byte[]>> results = new ArrayList<>(gets.size());
+      int hbaseResultsIndex = 0;
+      for (co.cask.cdap.api.dataset.table.Get get : gets) {
+        List<byte[]> cols = get.getColumns();
+        if (cols != null && cols.isEmpty()) {
```

Comment:
for code readability, I personally would have the same condition as in the the previous loop. I know this is the inverse condition, but whoever reads this code, needs to think twice to see that.